File: Java/PdfViewerAPI_R2/java/com/pdf_tools/pdfviewer/DocumentManagement/IPdfDocument.java
/***************************************************************************
*
* File:            IPdfDocument.java
* 
* Package:         3-Heights(TM) PDF Java Viewer
*
* Description:     The interface of a PDF document.
*
* @author          Christian Hagedorn, PDF Tools AG   
* 
* Copyright:       Copyright (C) 2001 - 2008 PDF Tools AG, Switzerland
*                  All rights reserved.
*                  
***************************************************************************/

package com.pdf_tools.pdfviewer.DocumentManagement;

import java.awt.image.BufferedImage;
import java.util.List;
import java.util.Map;

import com.pdf_tools.pdfviewer.Model.PdfAdvancedDestination;
import com.pdf_tools.pdfviewer.Model.PdfCanvas;
import com.pdf_tools.pdfviewer.Model.PdfOutlineItem;
import com.pdf_tools.pdfviewer.Model.PdfTextFragment;
import com.pdf_tools.pdfviewer.Model.PdfViewerException;
import com.pdf_tools.pdfviewer.Annotations.APdfAnnotation;
import com.pdf_tools.pdfviewer.Annotations.PdfGenericAnnotation;
import com.pdf_tools.pdfviewer.Annotations.APdfAnnotation.TPdfAnnotationType;
import com.pdf_tools.pdfviewer.Model.IPdfCommon.TPageLayoutMode;
import com.pdf_tools.pdfviewer.Model.PdfViewerController.Viewport;
import com.pdf_tools.pdfviewer.Model.PdfViewerException.NoFileOpenedException;
import com.pdf_tools.pdfviewer.converter.geom.Rectangle;

/**
 * The IPdfDocument interface describes the interface to the document class.
 * 
 * @author cha
 *
 */
public interface IPdfDocument
{
    /**
     * Open the file with the associated password located at the absolute path
     * specified by filename. If the file does not contain a password an empty
     * string must be passed. This method is invoked by {@link PdfCanvas#open()}
     * method of {@code PdfCanvas} class in the delegation chain.
     * 
     * @param filename
     *            The absolute path to the file
     * @param password
     *            The password needed to open the file (if no password required
     *            pass "" as password)
     * @throws PdfViewerException
     */
    public void open(String filename, String password) throws PdfViewerException;

    /**
     * Close the currently open document. This method is invoked by
     * {@link PdfCanvas#close()} method of {@code PdfCanvas} class in the
     * delegation chain.
     * 
     * @throws PdfViewerException
     */
    public void close() throws PdfViewerException;

    /**
     * Save the current file to path
     * 
     * @param path
     * @throws PdfViewerException
     */
    public void saveAs(String path) throws PdfViewerException;

    /**
     * Create new annotation
     * 
     * @param type:
     *            Type of the annotation
     * @paran page: page of the annotation
     * @param rect:
     *            Rectangle of the annotation
     * @returns PdfGenericAnnotation the newly created annotation that has to be
     *          added to the annotation cache
     * @throws PdfViewerException
     */
    public PdfGenericAnnotation createAnnotation(TPdfAnnotationType type, int page, Double[] rect) throws PdfViewerException;

    /**
     * Delete annotation
     * 
     * @param annotHandle:
     *            pointer to the annotation
     * @returns boolean: true for successful deletion, false otherwise
     */

    public void deleteAnnotation(long annotHandle, int page) throws PdfViewerException;

    /**
     * Returns the number of pages of the currently open document. If the
     * current document is not open then 0 is returned.
     * 
     * @return The number of pages
     * @throws com.pdf_tools.pdfviewer.Model.PdfViewerException.NoFileOpenedException
     */
    public int getPageCount() throws NoFileOpenedException;

    public Map<Integer, Rectangle.Double> getPageRangeFromSource(List<Integer> pages) throws PdfViewerException;

    public boolean isOpen();

    public TPageLayoutMode getPageLayout() throws PdfViewerException;

    public PdfAdvancedDestination getOpenActionDestination() throws PdfViewerException;

    public PdfOutlineItem[] getOutlines(int parentId) throws PdfViewerException;

    public PdfGenericAnnotation[] getAnnotationsOnPage(int pageNo) throws PdfViewerException;

    public void updateAnnotation(APdfAnnotation annotation) throws PdfViewerException;

    public List<PdfTextFragment> getTextFragments(int pageNo) throws PdfViewerException;

    public void draw(BufferedImage bitmap, int rotation, Map<Integer, Rectangle.Double> pageRectsDict, Viewport viewport)
            throws PdfViewerException;

    BufferedImage LoadThumbnail(double sourceWidth, double sourceHeight, int targetWidth, int targetHeight, int page)
            throws PdfViewerException;

}

File: CS.NET/PdfViewerCSharpAPI_R2/DocumentManagement/IPdfDocument.cs
/****************************************************************************
 *
 * File:            IPdfDocumens.cs
 *
 * Description:     The 3-Heights™  Pdf Viewer Control for WPF
 *
 * Copyright:       Copyright (C) 2015 PDF Tools AG, Switzerland
 *                  All rights reserved.
 *                  
 ***************************************************************************/


using PdfTools.PdfViewerCSharpAPI.Annotations;

namespace PdfTools.PdfViewerCSharpAPI.DocumentManagement
{
    using System;
    using System.Collections.Generic;
    using System.Collections.Concurrent;
    using System.Linq;
    using System.Text;
    using System.Windows.Media.Imaging;
    using PdfTools.PdfViewerCSharpAPI.Utilities;
    using PdfTools.PdfViewerCSharpAPI.Model;
    using PdfTools.PdfViewerCSharpAPI.DocumentManagement;

    /// <summary>
    /// Interface of a Document, which represents a pdf file in its raw form
    /// </summary>
    public interface IPdfDocument : IDisposable
    {
        /// <summary>
        /// Open a new file. This document will from now on represent this file
        /// </summary>
        /// <param name="filename"></param>
        /// <param name="password"></param>
        void Open(string filename, byte[] fileMem, string password);
       
        /// <summary>
        /// Closes the opened file. This document will not represent anything until a new file is opened
        /// </summary>
        void Close();


        IList<PdfTextFragment> LoadTextFragments(int pageNo);

        /// <summary>
        /// Get the amount of pages in this document
        /// </summary>
        int PageCount
        {
            get;
        }


        PdfSourceRect GetPageRect(int pageNo);

        void Draw(WriteableBitmap bitmap, int rotation, IList<KeyValuePair<int, PdfSourceRect>> pageRects, PdfViewerController.Viewport viewport);

        WriteableBitmap LoadThumbnail(double sourceWidth, double sourceHeight, int targetWidth, int targetHeight, int page, Resolution resolution);

        TPageLayoutMode GetPageLayout();
        PdfDestination GetOpenActionDestination();
        IList<PdfOutlineItem> GetOutlines(int parentId);

        #region [InkingForPDF] Annotation Methods

        /// <summary>
        /// Returns a List of all Annotations on a page
        /// </summary>
        /// <param name="pageNr"></param>
        /// <returns></returns>
        IList<PdfAnnotation> GetAnnotations(int pageNr);

        /// <summary>
        /// Creates a new Annotation
        /// </summary>
        /// <param name="annotType"></param>
        /// <param name="pageNr"></param>
        /// <param name="annotPoints"></param>
        /// <param name="annotPointsLength"></param>
        /// <param name="color"></param>
        /// <param name="colorLength"></param>
        /// <param name="strokeWidth"></param>
        /// <returns></returns>
        IntPtr CreateAnnotation(PdfDocument.TPdfAnnotationType annotType, int pageNr, double[] annotPoints, int annotPointsLength, double[] color, int colorLength, double strokeWidth);

        /// <summary>
        /// Retrieves pointer and size of the annotation array
        /// </summary>
        /// <param name="pageNr"></param>
        /// <param name="annotsPointer"></param>
        /// <param name="annotsLength"></param>
        /// <returns></returns>
        bool LoadAnnotations(int pageNr, out IntPtr annotsPointer, out int annotsLength);

        /// <summary>
        /// Updates an annotation with new values
        /// </summary>
        /// <param name="annotId"></param>
        /// <param name="pageNr"></param>
        /// <param name="boundingBox"></param>
        /// <param name="content"></param>
        /// <param name="label"></param>
        /// <param name="color"></param>
        /// <param name="strokeWidth"></param>
        /// <returns></returns>
        int UpdateAnnotation(IntPtr annotId, int pageNr, double[] boundingBox, string content, string label, double[] color, double strokeWidth);

        /// <summary>
        /// Deletes an annotation with the given id
        /// </summary>
        /// <param name="annotId"></param>
        void DeleteAnnotation(IntPtr annotId);

        /// <summary>
        /// Saves the document at the given path
        /// </summary>
        /// <param name="filePath"></param>
        /// <returns></returns>
        bool SaveAs(string filePath);

        #endregion [InkingForPDF] Annotation Methods
    }
}
